### PR TITLE
Subpartitioning: Refactors code to replace internal flags with preview flags for public preview

### PR DESCRIPTION
**Pull request by SrinikhilReddy**
# Enable subpartitioning feature to be available as public preview.

## Description
Replacing the existing internal preprocessor directives with preview for subpartitioning feature code.
```
## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 1]
```diff
@@ -99,7 +99,11 @@ public class CosmosClient : IDisposable
 
         static CosmosClient()
         {
+            #if PREVIEW
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:102]: Please run the ./UpdateContracts.ps1 to update the preview contract file.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:101]: This is no longer INTERNAL, we should remove the Internal keyword from the PR

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 2]
```diff
@@ -98,7 +98,11 @@ public class CosmosClient : IDisposable
...
+            HttpConstants.Versions.CurrentVersion = HttpConstants.Versions.v2020_07_15;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:102]: Are there any other changes we are bringing with the version bump from the protocol/wire perspective?

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 1 above)
Comment (SrinikhilReddy) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:101]: Removed.

**Comment by j82w** [comment]
Closing since there has been no update for over 30 days. Please re-open if necessary.

**Comment by SrinikhilReddy** [comment]
This PR was closed because of a lack of Java support for subpartitioning. Java PR is now out, https://github.com/Azure/azure-sdk-for-java/pull/18503. Reopening this PR for merge consideration.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 2 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:103]: Had similar comments on JAVA version.

File: Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace Microsoft.Azure.Cosmos
     /// <summary>
     /// Represents a partition key value list in the Azure Cosmos DB service.
     /// </summary>
-#if INTERNAL || SUBPARTITIONING
+#if PREVIEW
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/PartitionKeyBuilder.cs:14]: Please cover in the code docs that its an ordered pk collection

File: Microsoft.Azure.Cosmos/src/Resource/Database/DatabaseCore.cs [hunk 4]
```diff
@@ -228,7 +228,7 @@ internal abstract class DatabaseCore : DatabaseInternal
                                 nameof(containerProperties.PartitionKey));
                         }
                     }
-#if INTERNAL || SUBPARTITIONING
+#if PREVIEW
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Database/DatabaseCore.cs:231]: @j82w, @ealsur can this branch be made unconditional?
Or what's the use-case to make it conditional, if the top level composition is not available.

**Review by kirankumarkolli — CHANGES_REQUESTED** [review]
Lets please ensure that JAVA and .NET API's are in sync. 
JAVA has new comments and if deviates then similar changes are needed in .NET

File: Microsoft.Azure.Cosmos/src/Resource/Database/DatabaseCore.cs (hunk 4 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Database/DatabaseCore.cs:231]: Yes, it was added to make it easier to produce a private nuget with the feature enabled.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 2 above)
Comment (SrinikhilReddy) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:103]: Hi @ealsur, @kirankumarkolli The version bump helps support the subpartitioning change. Currently subpartitiong and merge are the features supported on this version

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:103]: Sub-part version is just after current version. Not an issue here.

**Review by kirankumarkolli — DISMISSED** [review]

**Review by j82w — APPROVED** [review]

**Pull request merged**